File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        
        def possible(cap):
            total, day = cap, 1
            
            for i in weights:
                if total < i:
                    day, total = day+1, cap - i
                else:
                    total -= i
                    
            return True if day <= days else False
            
        left, right, ans = max(weights), sum(weights), 0
        
        while left <= right:
            mid = (left + right) // 2
            
            if possible(mid):
                ans, right = mid, mid - 1
            else:
                left = mid + 1
                    
        return ans

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        function<bool(int)> f = [&] (long long x) {
            int cur = 0, cnt = 0;
            for (int i = 0; i < weights.size(); i++) {
                if (weights[i] > x)
                    return false;
                if (cur + weights[i] > x) {
                    cur = 0;
                    cnt++;
                }
                cur += weights[i];
            }
            cnt += (cur > 0);
            return cnt <= days;
        };
        
       long long lo = 0, hi = accumulate(weights.begin(), weights.end(), 0ll);
        while (lo < hi) {
            long long mid = lo + (hi - lo) / 2;
            if (f(mid))
                hi = mid;
            else
                lo = mid + 1;
        }
        return lo;
    }
};